src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -84,5 +224,187 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
                     .ToList();
             }
         }
+
+        /// <summary>
+        /// Get All Installed packages that transitively install a given transitive package in this project
+        /// </summary>
+        /// <param name="transitivePackage">Identity of given transtive package</param>
+        /// <param name="ct">Cancellation Token</param>
+        /// <returns>A dictionary, indexed by Framework/Runtime-ID with all top (installed)
+        /// packages that depends on given transitive package, or <c>null</c> if none found</returns>
+        /// <remarks>Computes all transitive origings for each Framework/Runtime-ID combiation. Runtime-ID can be <c>null</c>.
+        /// Transitive origins are calculated using a Depth First Search algorithm on all direct dependencies exhaustively</remarks>
+        internal TransitiveEntry GetTransitivePackageOrigin(PackageReference transitivePackage, IReadOnlyList<PackageReference> installedPackages, IList<LockFileTarget> targetsList, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+            if (IsInstalledAndTransitiveComputationNeeded)
+            {
+                // Assets file has changed. Recompute Transitive Origins
+                TransitiveOriginsCache.Clear();
+
+                // Otherwise, find all Transitive origin and update cache
+                var memoryVisited = new HashSet<PackageIdentity>();
+
+                // 3. For each target framework graph (Framework, RID)-pair:
+                foreach (LockFileTarget targetFxGraph in targetsList)
+                {
+                    var key = new FrameworkRuntimePair(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
+
+                    foreach (var directPkg in installedPackages) // 3.1 For each direct dependency d:
+                    {
+                        memoryVisited.Clear();
+                        // 3.1.1 Do DFS to mark directPkg as a transitive origin over all transitive dependencies found
+                        MarkTransitiveOrigin(directPkg, directPkg, targetFxGraph, memoryVisited, key, ct);
+                    }
+                }
+            }
+            // Otherwise, assets file has not changed. Look at Transitive Origins Cache
+            TransitiveEntry cacheEntry;
+            TransitiveOriginsCache.TryGetValue(transitivePackage.PackageIdentity.Id, out cacheEntry);
```

Comment:
If 2 versions of same package Id (i.e newsoft.json) may have different origin path tree. Are we accounting that here? `Id_version` could make better key.

Reply:
Issue is that dependencies section contains the _required_ version, which can be different from _resolved_ version.